text/3349-cfg-os-version.md

```diff
@@ -0,0 +1,142 @@
+
+By default, the standard library uses a sensible minimum API version.  For example, for `x86_64-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7.  However, there are good reasons why you might want control of how your code is compiled depending on what the minimum API version is set as.  For instance, if you want to:
+
+* set your crate's minimum API version higher than that of the standard library.
+* change certain implementation details of your crate depending on what a downstream user sets their minimum API version to be.
+* have some sensible compiler error if users of your crate require a lower minimum API version that you require.
+
+In cases like these you can use the `os_version_min` and `os_version_range` predicates to specify the minimum API levels of various parts of the operating system.  For example:
+
+* `os_version_min(“windows”, <string>)` would test the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `os_version_min(“libc”, <string>)` would test the version of libc in use.
+* `os_version_min(“kernel”, <string>)` would test the version of the kernel in use.
+
+Let’s use `os_version_min(“windows”, …)` as an example.  It should be clear how this example would be extended to the other `cfg` predicates. The predicate allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(os_version_min(“windows”, "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+Crate authors can set the API requirements of their Cargo configuration file under the  [`target key`](https://doc.rust-lang.org/cargo/reference/config.html#target) like so (suggested variable name/syntax only):
+
+```toml
+[target.x86_64-pc-windows-msvc]
+min_os_version_windows = "6.0.6000" # Vista
+```
+
+When compiling, the user can provide the API levels to compile for: `rustc --cfg 'target_os_version.windows="6.0.6000"'`.
+
+If an end user sets their `os_version.windows` to an incompatible version then the user receives an error. For instance, in the example above where the user is setting their `min_os_version_windows` to Windows Vista, they will receive an error when linking with the standard library which imposes Windows 7 as its minimum `os_version.windows` by default for the `x86_64-pc-windows-msvc` target.
```

Comment:
I assume this is what is meant here?

```suggestion
If an end user sets their `target_os_version.windows` to an incompatible version then the user receives an error. For instance, in the example above where the user is setting their `min_os_version_windows` to Windows Vista, they will receive an error when linking with the standard library which imposes Windows 7 as its minimum `target_os_version.windows` by default for the `x86_64-pc-windows-msvc` target.
```

Also, I'm a little unclear here, won't this make it impractical to set any min version above the standard library? I don't see a situation where setting this minimum could be used. (I'm likely having a very fundamental misunderstanding here, but I'm a bit confused.)  If this is *only* useful with build-std, it might be good to make that clear up-front.

Reply:
You are correct that you couldn't use a minimum version above that used by the standard library, but different versions of the standard library can be built with different minimum versions.  For example, the Android Rust toolchain uses a minimum Android API of 31 as opposed to API 19 used by the official Rust toolchain.

I believe the issue you are concerned about is mostly limited to "whole crate" minimum target versions, which I suspect (with no data to back it up) will make up the minority of use cases.  I would expect this `cfg` option to be used to conditionally compile individual bits of code instead.

Reply:
(Typo fixes applied to my local edits to this commit.)

Reply:
> won't this make it impractical to set any min version above the standard library?

To expand on what @chriswailes said, this can also be used in a `no_std` context as well. You are correct that this becomes much more useful in a `buildstd` context like in the aforementioned internal Android toolchain, but it's not *exclusively* useful there. 

As Chris also points out, libraries can use this to more directly indicate that they are reliant on an API that is present only in some minimum API version number. This is not currently possible to do unless you have some bespoke features indicating this which if used incorrectly lead to just bad linker errors.

Reply:
Usually `no_std` is equivalent to "no operating system", so it seems like `target_os_version` wouldn't really apply there? What would that mean?

If this is only for users that build the standard library themselves, I would suggest making that clear up front. That is generally not the primary workflow for most Rust developers (I suspect), so this seems like a relatively niche feature.

It also makes me wonder if this should be more hidden or otherwise not exposed in Cargo. For example, it sounds like it would not be usable for the typical user of crates.io.

Or maybe I am still quite confused as to how this would work.

Reply:
> Usually no_std is equivalent to "no operating system", so it seems like target_os_version wouldn't really apply there? What would that mean?

Some libraries might choose to be able to work in environments where std is not supported but that still have some operating system facilities. After all `no_std` does not mean no OS, just that std makes assumptions about the platform that the program cannot make. 

For examples, `windows-rs` has (at least last time I checked), `no_std` support even though its purpose is to interact with the Windows operating system. Users of that library sometimes cannot rely on the standard library because they are in a Windows context that does not support all the features that std relies on.

While the assumption of there not being an OS at all is certainly more common, it's not the only use of `no_std`.

> If this is only for users that build the standard library themselves, I would suggest making that clear up front

I think this would make sense to do with the caveat above of the possibility of this making sense in a `no_std` context. The most likely scenario for using this feature does seem to ultimately be in a `buildstd` environment.